### PR TITLE
E2E testing

### PR DESCRIPTION
**Pull request by biern**
Well this involved **much** more work than I anticipated, but the initial setup is finally done.

- CI runs e2e tests on Chrome
- The actual build generated in the first step is used in e2e testing to make the process more robust and less time consuming
- Tests are runnable locally via npm or docker as well. See readme for that.
- The initial tests are pretty basic, just check if after uploading a file its hash and filename appear. They will get more sophisticated later on.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -11,11 +11,12 @@
   ],
   "scripts": {
     "postinstall": "bower cache clean && bower --allow-root install",
-    "clean": "rimraf static/dist",
+    "clean": "rm -rf static/dist/*",
```
Comment (jksf) [inline package.json]: Why do you replace `rimraf` with `rm -rf`? It is not a priority but it breaks the build on Windows OS.

File: README.md [hunk 2]
```diff
@@ -34,9 +34,31 @@ hot-reloads your application when sources changes.
...
+#### Running end-to-end tests
+
+##### Running locally
+
+Runs on chrome by default.
+
+``` shell
+cd tests/e2e
+npm install
+npm run test
```
Comment (jksf) [inline README.md]: `npm test` will suffice

**Review by jksf — COMMENTED** [review]
I don't like the fact that we have two directories `test` and `tests` at the root of the project. Can we merge them?

**Comment by jksf** [comment]
I wonder if we can have a single root for npm. Running separate `npm install` in `tests/e2e` directory is cumbersome. Can we add `tests/e2e/package.json` dependencies to `dev-dependencies` in the root project and remove package nesting?

File: tests/e2e/tests/basic.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+test('Site is available', async t => {
+  await t
+    .expect(Selector('body').textContent)
+    .contains('SignHash');
+});
+
+
+test('Verifying known hash', async t => {
+  const file = knownFile;
+  const text = Selector('body').textContent
```
Comment (jksf) [inline tests/e2e/tests/basic.ts]: Missing semicolon at the end of the line

File: tests/e2e/tests/basic.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+  await t
+    .setFilesToUpload('#file-upload', file.path)
+    .expect(text).contains(file.name)
+    .expect(text).contains(file.hash)
+    ;
```
Comment (jksf) [inline tests/e2e/tests/basic.ts]: I'd rather have a semicolon at the end of the previous line. This is a matter of preferred style, and I know that it makes `git diff` better, but it is unusual.

**Comment by jksf** [comment]
To summarize, I appreciate this PR greatly :+1:. Issues that I commented are minor. I can make these changes myself if you are too busy.

File: package.json (hunk 1 above)
Comment (biern) [reply package.json]: This was one of many issue when running those tests on docker. `static/dist` was a mounted volume and couldn't be removed. You're right about breaking compatibility, I didn't think about that. I'll try to solve that using `rimraf`

**Comment by biern** [comment]
> I wonder if we can have a single root for npm. Running separate npm install in tests/e2e directory is cumbersome. Can we add tests/e2e/package.json dependencies to dev-dependencies in the root project and remove package nesting?

I don't think that's a good idea. It will only create unnecessary dependencies between testing framework and the app itself that might eventually create further issues. I consider those a separate projects and would like to keep them that way.

**Comment by biern** [comment]
> To summarize, I appreciate this PR greatly 👍. Issues that I commented are minor. I can make these changes myself if you are too busy.

Thanks! I can fix the remaining issues, but it probably won't happen soon as my schedule for the whole next week is full. In the meantime if you feel like working on that, feel free.

File: tests/e2e/tests/basic.ts (hunk 3 above)
Comment (biern) [reply tests/e2e/tests/basic.ts]: Weird that my linter didn't caught that.

File: tests/e2e/tests/basic.ts (hunk 4 above)
Comment (biern) [reply tests/e2e/tests/basic.ts]: Yeah, it's unusual. I don't have any strong opinion on that, it just felt more suitable to this code. It's not anything important at this point so, whatever works.

**Comment by biern** [comment]
I've fixed the remaining issues. I'm keeping the e2e tests as a separate npm project for now, we'll see how it performs.

**Pull request merged**